test/unit_poly1305_test.c

```diff
@@ -0,0 +1,552 @@
+/*
+ * Copyright 1999-2016 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/* Unit tests for the poly1305 module */
+/* This program must be explicitelly linked with the object files it tests,
+ * along with libcrypto (for all other functionality) */
+
+#include <stdio.h>
+#include <string.h>
+
+#include "testutil.h"
+#include "internal/poly1305.h"
+#include "../crypto/poly1305/poly1305_local.h"
+#include "e_os.h"
+
+typedef struct {
+    const char *inputhex;
+    const char *keyhex;
+    const char *outhex;
+} TESTDATA;
+
+typedef struct {
+    const char *test_case_name;
+    int test_num;
+    const TESTDATA *test_data;
+} SIMPLE_FIXTURE;
+
+/**********************************************************************
+ *
+ * Test of poly1305 internal functions
+ *
+ ***/
+
+static SIMPLE_FIXTURE setup_poly1305(const char *const test_case_name)
+{
+    SIMPLE_FIXTURE fixture;
+    fixture.test_case_name = test_case_name;
+    return fixture;
+}
+
+static unsigned char hex_digit(char h)
+{
+    int i = OPENSSL_hexchar2int(h);
+
+    if (i < 0)
+        abort();
+    return i;
+}
+
+static void hex_decode(unsigned char *out, const char *hex)
```

Comment:
Almost every test needs hex helpers so I think these should be cleaned up and moved to a common header (`testutil.h`?). However, I know you're just moving code around and I don't want to make you do the extra work so feel free to stick a `TODO(emilia)` in there. There's a bunch of duplication here that I plan to work on, anyhow.